admin/lib/solidus_admin/component_registry.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module SolidusAdmin
+  class ComponentRegistry
+    ComponentNotFoundError = Class.new(NameError)
+
+    def initialize
+      @names = {}
+    end
+
+    def []=(key, value)
+      @names[key] = value
+    end
+
+    def [](key)
+      if @names[key]
+        @names[key].constantize
+      else
+        infer_constant_from(key)
```

Comment:
We could store the constant in something like `@components[key]`, no? 

That way we would not constantize often used components (like buttons) several times per page.

```suggestion
        @components[key] = infer_constant_from(key)
```

Constantize is painfully slow and since this is called in a view context and during the hot path of rendering a page I think we should cache the component constants. 

```
Benchmark.ips do |x|
  x.report("read constant") { Alchemy }
  x.report("constantize") { "Alchemy".constantize }
  x.compare!
end
ruby 3.3.1 (2024-04-23 revision c56cd86388) [arm64-darwin23]
Warming up --------------------------------------
       read constant     3.749M i/100ms
         constantize   781.449k i/100ms
Calculating -------------------------------------
       read constant     37.175M (± 0.8%) i/s -    187.459M in   5.042887s
         constantize      7.762M (± 2.4%) i/s -     39.072M in   5.037306s

Comparison:
       read constant: 37175322.4 i/s
         constantize:  7761993.2 i/s - 4.79x  slower
```

Another idea would be to pre-register all components we have and only lazy load additional components, but this can be cumbersome.

Reply:
I don't think something that does almost eight million iterations in a second is slow enough to warrant the extra complexity. That's a lot of buttons.